dartagnan/src/main/java/com/dat3m/dartagnan/wmm/axiom/Acyclic.java

```diff
@@ -304,32 +277,36 @@ private BooleanFormula consistentSAT(Set<Tuple> toBeEncoded, Context analysisCon
             Tuple t3 = new Tuple(tri[0], tri[2]);
 
             BooleanFormula cond = minSet.contains(t3) ?
-                    execution(t3.getFirst(), t3.getSecond(), exec, ctx)
-                    : bmgr.and(getSMTCycleVar(t1, ctx), getSMTCycleVar(t2, ctx));
+                    context.execution(t3.getFirst(), t3.getSecond())
+                    : bmgr.and(getSMTCycleVar(t1, fmgr), getSMTCycleVar(t2, fmgr));
 
-            enc = bmgr.and(enc, bmgr.implication(cond, getSMTCycleVar(t3, ctx)));
+            enc = bmgr.and(enc, bmgr.implication(cond, getSMTCycleVar(t3, fmgr)));
         }
 
         //  --- Encode inconsistent assignments ---
         // Handle self-loops
         for (Event e : selfloops) {
-            enc = bmgr.and(enc, bmgr.not(rel.getSMTVar(e, e, ctx)));
+            enc = bmgr.and(enc, bmgr.not(context.edge(rel, new Tuple(e, e))));
         }
         // Handle remaining cycles
         for (int i = 0; i < varOrderings.size(); i++) {
             Set<Tuple> out = vertEleOutEdges.get(varOrderings.get(i));
             for (Tuple t : out) {
                 if (varOrderings.indexOf(t.getSecond()) > i && vertEleInEdges.get(t.getSecond()).contains(t)) {
-                    BooleanFormula cond = minSet.contains(t) ? bmgr.makeTrue() : getSMTCycleVar(t, ctx);
-                    enc = bmgr.and(enc, bmgr.implication(cond, bmgr.not(getSMTCycleVar(t.getInverse(), ctx))));
+                    BooleanFormula cond = minSet.contains(t) ? bmgr.makeTrue() : getSMTCycleVar(t, fmgr);
+                    enc = bmgr.and(enc, bmgr.implication(cond, bmgr.not(getSMTCycleVar(t.getInverse(), fmgr))));
                 }
             }
         }
 
         return enc;
     }
 
-    private BooleanFormula getSMTCycleVar(Tuple edge, SolverContext ctx) {
-        return edge(getName() + "-cycle", edge.getFirst(), edge.getSecond(), ctx);
+    private BooleanFormula cycleVar(Event event, FormulaManager m) {
+        return m.getBooleanFormulaManager().makeVariable(String.format("cycle %s %d", m.escape(getName()), event.getCId()));
+    }
+
+    private BooleanFormula getSMTCycleVar(Tuple edge, FormulaManager m) {
+        return m.getBooleanFormulaManager().makeVariable(String.format("cycle %s %d %d", m.escape(getName()), edge.getFirst().getCId(), edge.getSecond().getCId()));
     }
```

Comment:
It seems the `BooleanFormulaManager` is available in every call to these methods. Why not to pass the `BooleanFormulaManager` instead of `FormulaManager` directly?

Reply:
Because I use `escape` here.

Reply:
I could assert `isValidName` instead.